milli/src/documents/mod.rs

```diff
@@ -0,0 +1,231 @@
+mod builder;
+/// The documents module defines an intermediary document format that milli uses for indexation, and
+/// provides an API to easily build and read such documents.
+///
+/// The `DocumentBuilder` interface allows to write batches of documents to a writer, that can
+/// later be read by milli using the `DocumentsReader` interface.
+mod reader;
+mod serde;
+
+use std::{fmt, io};
+
+use ::serde::{Deserialize, Serialize};
+use bimap::BiHashMap;
+pub use builder::DocumentsBuilder;
+pub use reader::DocumentsReader;
+
+use crate::FieldId;
+
+#[derive(Debug, Serialize, Deserialize)]
+struct DocumentsMetadata {
+    count: usize,
+    index: BiHashMap<FieldId, String>,
+}
+
+pub struct ByteCounter<W> {
+    count: usize,
+    writer: W,
+}
+
+impl<W> ByteCounter<W> {
+    fn new(writer: W) -> Self {
+        Self { count: 0, writer }
+    }
+}
+
+impl<W: io::Write> io::Write for ByteCounter<W> {
+    fn write(&mut self, buf: &[u8]) -> io::Result<usize> {
+        let count = self.writer.write(buf)?;
+        self.count += count;
+        Ok(count)
+    }
+
+    fn flush(&mut self) -> io::Result<()> {
+        self.writer.flush()
+    }
+}
+
+#[derive(Debug)]
+pub enum Error {
+    InvalidDocumentFormat,
+    Custom(String),
+    JsonError(serde_json::Error),
+    Serialize(bincode::Error),
+    Io(io::Error),
+    DocumentTooLarge,
+}
+
+impl From<io::Error> for Error {
+    fn from(other: io::Error) -> Self {
+        Self::Io(other)
+    }
+}
+
+impl From<bincode::Error> for Error {
+    fn from(other: bincode::Error) -> Self {
+        Self::Serialize(other)
+    }
+}
+
+impl fmt::Display for Error {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        match self {
+            Error::Custom(s) => write!(f, "Unexpected serialization error: {}", s),
+            Error::InvalidDocumentFormat => f.write_str("Invalid document addition format."),
+            Error::JsonError(err) => write!(f, "Couldn't serialize document value: {}", err),
+            Error::Io(e) => e.fmt(f),
+            Error::DocumentTooLarge => f.write_str("Provided document is too large (>2Gib)"),
+            Error::Serialize(e) => e.fmt(f),
+        }
+    }
+}
+
+impl std::error::Error for Error {}
+
+/// Macro used to generate documents, with the same syntax as `serde_json::json`
+#[cfg(test)]
+macro_rules! documents {
+    ($data:tt) => {{
+        let documents = serde_json::json!($data);
+        let mut writer = std::io::Cursor::new(Vec::new());
+        let mut builder =
+            crate::documents::DocumentsBuilder::new(&mut writer, bimap::BiHashMap::new()).unwrap();
```

Comment:
For example, here we should not use `&mut writer` but rather `writer` directly and call `into_inner` to get the `writer` back without calling `finish` on it.

Reply:
yes but nothing prevents you from doing it, so it's not a strong guarantee...

Reply:
If you don't call `into_inner` here you will not be able to retrieve your `writer` and write it in the obkv, so yeah it prevents you from forgetting to call `into_inner` (where we flush on disk).